### PR TITLE
[FAB-17709] Support creating system channel genesis block

### PR DESCRIPTION
**Pull request by wuxuer**
Add a function to create a system channel genesis block.

Signed-off-by: xu wu <wuxu1103@163.com>

#### Type of change

- New feature

#### Related issues
https://jira.hyperledger.org/browse/FAB-17709

### CODE REVIEW

**Review by DereckLuo — DISMISSED** [review]
+1 LGTM

File: pkg/config/config.go [hunk 1]
```diff
@@ -210,6 +212,28 @@ func NewCreateChannelTx(channelConfig Channel, channelID string) (*cb.Envelope,
...
+	}
+
+	ct, err := newSystemChannelGroup(channelConfig)
+	if err != nil {
+		return nil, fmt.Errorf("creating system channel group: %v", err)
+	}
+
+	block, err := newSystemChannelBlock(ct, channelID)
+	if err != nil {
+		return nil, fmt.Errorf("creating system channel block: %v", err)
```
Comment (caod123) [inline pkg/config/config.go]: likewise to my comment on the function name, I think this error will need some rework too. We probably want to focus more on the fact that this is the genesis block as otherwise `creating system channel block` can be misconstrued for what this function is intended to be used for.

File: pkg/config/config.go [hunk 2]
```diff
@@ -210,6 +212,28 @@ func NewCreateChannelTx(channelConfig Channel, channelID string) (*cb.Envelope,
 	return env, nil
 }
 
+// NewSystemChannelBlock creates a genesis block using the provided consortiums and orderer
+// configuration and returns a block.
+func NewSystemChannelBlock(channelConfig Channel, channelID string) (*cb.Block, error) {
```
[the remark this answers is not in the document]
Comment (caod123) [reply pkg/config/config.go]: the most clear API function would be `NewSystemChannelGenesisBlock` but that might be a bit wordy

File: pkg/config/config.go [hunk 3]
```diff
@@ -210,6 +212,28 @@ func NewCreateChannelTx(channelConfig Channel, channelID string) (*cb.Envelope,
...
+	var err error
+
+	if channelID == "" {
+		return nil, errors.New("system channel ID is required")
+	}
+
+	ct, err := newSystemChannelGroup(channelConfig)
```
Comment (caod123) [inline pkg/config/config.go]: you may want to update this function and the `newChannelGroup` function for application channels. Application channels must have the `channelConfig.Consortium` specified whereas system channels do not (looks like this behavior on master is currently swapped). `newSystemChannelGroup` and `newChannelGroup` are also both missing a modPolicy: `channelGroup.ModPolicy = channelConfig.AdminsPolicyKey`

**Review by caod123 — CHANGES_REQUESTED** [review]

File: pkg/config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wuxuer) [reply pkg/config/config.go]: In order to clarify the meaning of the function, I changed the function to be NewSystemChannelGenesisBlock .

File: pkg/config/config.go (hunk 1 above)
Comment (wuxuer) [reply pkg/config/config.go]: I also added "genesis" in the error.

File: pkg/config/config.go (hunk 3 above)
Comment (wuxuer) [reply pkg/config/config.go]: I fixed the two problems in fucntion newSystemChannelGroup and newChannelGroup. Please review whether it is correct.

File: pkg/config/config.go [hunk 4]
```diff
@@ -343,15 +369,6 @@ func newSystemChannelGroup(channelConfig Channel) (*cb.ConfigGroup, error) {
 		return nil, err
 	}
 
-	if channelConfig.Consortium == "" {
-		return nil, errors.New("consortium is not defined in channel config")
-	}
```
Comment (caod123) [inline pkg/config/config.go]: This check probably should move to `newChannelGroup`

File: pkg/config/config_test.go [hunk 5]
```diff
@@ -409,6 +410,741 @@ func TestNewCreateChannelTxFailure(t *testing.T) {
 	}
 }
 
+func TestSystemChannelGenesisBlock(t *testing.T) {
```
Comment (caod123) [inline pkg/config/config_test.go]: nit: `TestNewSystemChannelGenesisBlock`

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -702,7 +1438,7 @@ func baseSystemChannelProfile(t *testing.T) Channel {
 		Orderer:      baseSoloOrderer(t),
 		Capabilities: []string{"V2_0"},
 		Policies:     standardPolicies(),
-		Consortium:   "testconsortium",
+		Consortium:   "",
```
Comment (caod123) [inline pkg/config/config_test.go]: You might as well remove this line entirely

File: pkg/config/config_test.go [hunk 7]
```diff
@@ -659,7 +1395,7 @@ func TestChannelConfiguration(t *testing.T) {
 				Orderer:      baseOrderer,
 				Capabilities: []string{"V2_0"},
 				Policies:     policies,
-				Consortium:   "testconsortium",
+				Consortium:   "",
```
Comment (caod123) [inline pkg/config/config_test.go]: You can probably remove this alongside line 1384

File: pkg/config/config_test.go [hunk 8]
```diff
@@ -409,6 +410,741 @@ func TestNewCreateChannelTxFailure(t *testing.T) {
...
+	actualChannelHeader := &cb.ChannelHeader{}
+	err = proto.Unmarshal(actualPayload.Header.ChannelHeader, actualChannelHeader)
+	gt.Expect(err).NotTo(HaveOccurred())
+	expectedChannelHeader.Timestamp = actualChannelHeader.Timestamp
+	expectedChannelHeader.TxId = actualChannelHeader.TxId
+
+	gt.Expect(actualChannelHeader).To(Equal(expectedChannelHeader))
+}
+
+func TestSystemChannelGenesisBlockFailure(t *testing.T) {
```
Comment (caod123) [inline pkg/config/config_test.go]: nit: `TestNewSystemChannelGenesisBlockFailure`

**Review by caod123 — CHANGES_REQUESTED** [review]
Some small requested changes

File: pkg/config/config.go (hunk 4 above)
Comment (wuxuer) [reply pkg/config/config.go]: Yes, I have fixed it and added an unit test case.

File: pkg/config/config_test.go (hunk 5 above)
Comment (wuxuer) [reply pkg/config/config_test.go]: Sorry for the following four nits. I have fixed them all.

File: pkg/config/config.go [hunk 9]
```diff
@@ -217,13 +242,17 @@ func newChannelGroup(channelConfig Channel) (*cb.ConfigGroup, error) {
 		if err != nil {
 			return nil, err
 		}
+	} else {
+		return nil, errors.New("consortium is not defined in channel config")
```
Comment (C0rWin) [inline pkg/config/config.go]: given you return within else, might make sense to revert logic of the `if-then-else` to save on indentation and return on error case immediately?

File: pkg/config/config.go [hunk 10]
```diff
@@ -364,6 +386,64 @@ func defaultConfigTemplate(channelConfig Channel) (*cb.ConfigGroup, error) {
...
+// newSystemChannelBlock generates a genesis block by the config group and system channel ID
+// the block num is always zero
+func newSystemChannelBlock(cg *cb.ConfigGroup, channelID string) (*cb.Block, error) {
+	payloadChannelHeader := channelHeader(cb.HeaderType_CONFIG, msgVersion, channelID, epoch)
+	nonce, err := newNonce()
+	if err != nil {
+		return nil, fmt.Errorf("try to get nonce: %v", err)
+	}
+	payloadSignatureHeader := &cb.SignatureHeader{Creator: nil, Nonce: nonce}
+	payloadChannelHeader.TxId = computeTxID(payloadSignatureHeader.Nonce, payloadSignatureHeader.Creator)
```
Comment (C0rWin) [inline pkg/config/config.go]: given that you generate `TxId` from fields of `SignatureHeader` doesn't it make sense to pass pointer to the struct itself rather than fields? also it's seems that creator is always a nil.

File: pkg/config/config.go [hunk 11]
```diff
@@ -364,6 +386,64 @@ func defaultConfigTemplate(channelConfig Channel) (*cb.ConfigGroup, error) {
...
+	}
+	envelope := &cb.Envelope{Payload: envelopePayload, Signature: nil}
+	blockData, err := proto.Marshal(envelope)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling envelope: %v", err)
+	}
+
+	block := newBlock(0, nil)
+	block.Data = &cb.BlockData{Data: [][]byte{blockData}}
+	block.Header.DataHash = blockDataHash(block.Data)
```
Comment (C0rWin) [inline pkg/config/config.go]: IMO it might make sense to encapsulate this initialization within `newBlock`

**Review by C0rWin — COMMENTED** [review]
thanks for your contribution I have a few small nits.

File: pkg/config/config.go (hunk 9 above)
Comment (wuxuer) [reply pkg/config/config.go]: Thank you very much for your suggestion. I have reverted the logic.

File: pkg/config/config.go (hunk 10 above)
Comment (wuxuer) [reply pkg/config/config.go]: Function computeTxID should be a general function. If the input parameter is *SignatureHeader, the versatility of this function will be greatly reduced. 
The design of function newBlock  is also based on the same logic, which constructs a block with no data and no metadata.
In addition, these two functions are referenced in the “protoutil/blockutils.go” file.
These are some of my thoughts, I don't know if they are suitable.

**Review by C0rWin — APPROVED** [review]

**Review by caod123 — APPROVED** [review]
